experimental/PlaneCurve/EllCurveZnZ.jl

```diff
@@ -0,0 +1,696 @@
+    if !isone(even)
+       !isone(S(D)) && (false, (ZZ(0), ZZ(0)))
+       s2 = sqrtmod(-D, even)
+       s = abs(crt(-s1, odd, -s2+ZZ(2)^(Int(log2(even)-1)), even, true))
+    else
+       s = s1
+    end
+
+    if 2*s > m
+        B = m - s
+    else
+	    B = s
+    end
+
+    A = m
+    while B^2 >= m
+       A, B = B, A%B
+    end
+    t = m - B^2
+    L = divrem(t, D)
+    if !iszero(L[2])
+       return (false, (ZZ(0), ZZ(0)))
+    elseif !issquare(L[1])
+        return (false, (ZZ(0), ZZ(0)))
+    else
+        return (true, [B, isqrt(L[1])])
+    end
+end
+
+################################################################################
+###############Hilbert Class Polynomial#########################################
+################################################################################
+
+#Compute reduced forms of imaginary quadratic number fields
+# based on Algo 5.3.5 in the book by Cohen
+function class_data(D::fmpz)
+   D<ZZ(0) || error("Input needs to be negative")
+
+   h = 0
+   b = abs(D % 2)
+   B = root(div(-D, 3), 2)
+   L = Vector{Tuple{fmpz,fmpz,fmpz}}()
+
+   while b <= B
+      q = div(b^2-D, 4)
+      a = ZZ(max(b, 1))
+      while a^2 <= q
+         if q % a == 0 && gcd(a,b, div(q,a)) == 1
+            if a == b || a^2 == q || b == 0
+               h = h+1
+               push!(L, (a,b, div(b^2-D, 4*a)))
+            else
+               h = h + 2
+               v = div(b^2-D, 4*a)
+               push!(L, (a,-b, v))
+               push!(L, (a,b, v))
+            end
+         end
+         a = a + 1
+      end
+      b = b + 2
+   end
+   return (ZZ(h), L)
+end
+
+
+function  hilbert_class_polynomial(R::FmpzPolyRing, d::Int)
+   z = R()
+   d < 0 && mod(d, 4) < 2 || error("oops")
+   ccall((:acb_modular_hilbert_class_poly, Nemo.libarb), Nothing,
+         (Ref{fmpz_poly}, Int),
+         z, d)
+   return z
+end
+
+#################################################################################
+
+function funddiscriminant(n::Int)
+       D = ZZ(-3)
+       L = Vector{fmpz}()
+       LL = Vector{Tuple{fmpz, fmpz}}()
+       while abs(D) <= n
+          if mod(D, 4) == 1 && issquarefree(D)
+             (h, _) = class_data(D)
+             push!(LL, (h, D))
+          end
+          if mod(D, 4) == 0 && issquarefree(div(D, 4)) && (mod(div(D,4),4) == 2 || mod(div(D,4), 4) == 3)
+             (h, _) = class_data(D)
+             push!(LL, (h, D))
+          end
+          D = D-1
+       end
+       LL = sort(LL, by = first)
+       i = 1
+       while LL[i][1] <= 4
+          push!(L, LL[i][2])
+          i = i + 1
+       end
+       return L
+end
+
+################################################################################
+# Miller Rabin Test
+################################################################################
+function _Miller_Rabin_witness(N::fmpz, a::fmpz, even::fmpz, odd::fmpz)
+	R = ResidueRing(ZZ, N)
+        x = ZZ(R(a)^odd)
+	if x == ZZ(1) || x == N - ZZ(1)
+		return false
```

Comment:
Hmm, this code and a bunch of the other is really oddly formatted. Perhaps consider using `JuliaFormatter` to get it a bit more uniform? Not that it's very important right now